### PR TITLE
Move mentions logic to status-go

### LINKED ISSUE #15220: Move mentions logic to status-go
Currently all mentions related logic in react, suggestions list, replacement in the text-input, and in the messages, we should move this to status-go

current implementation is here https://github.com/status-im/status-mobile/blob/develop/src/status_im/chat/models/mentions.cljs

### PR DESCRIPTION
**Pull request by qfrank**
close #15220 

relate PR for backend: [here](https://github.com/status-im/status-go/pull/3309)

status: ready

### CODE REVIEW

File: src/status_im/ui2/screens/chat/composer/input.cljs [hunk 1]
```diff
@@ -220,16 +217,15 @@
...
-         (partial on-change last-text-change timeout-id mentionable-users refs chat-id sending-image)
-         :on-text-input (partial on-text-input mentionable-users chat-id)}
+         (partial on-change last-text-change timeout-id refs chat-id sending-image)
+         :on-text-input (partial on-text-input chat-id)}
         input-with-mentions (rf/sub [:chat/input-with-mentions])
         children (fn []
                    (if mentions-enabled?
                      (map-indexed
-                      (fn [index [_ text]]
+                      (fn [index [type text]]
```
Comment (flexsurfer) [inline src/status_im/ui2/screens/chat/composer/input.cljs]: probably better to rename it so we don't have this issue again, something like `mention-type`

File: src/status_im/ui/screens/chat/components/input.cljs [hunk 2]
```diff
@@ -206,46 +203,20 @@
...
-        new-text             (string/join (map second hydrated-mentions))]
-    {:set-text-input-value [chat-id new-text]
-     :db
-     (-> db
-         (assoc-in [:chats/cursor chat-id] (:mention-end info))
-         (assoc-in [:chat/inputs-with-mentions chat-id] hydrated-mentions)
-         (assoc-in [:chats/mentions chat-id :mentions] info))}))
+  {:json-rpc/call [{:method     "wakuext_chatMentionToInputField"
+                    :params     [chat-id text]
+                    :on-success #(rf/dispatch [::mentions/on-to-input-field-success %])
```
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/components/input.cljs]: by guidelines we shouldn't use namespaced keywords

File: src/status_im/chat/models/input.cljs [hunk 3]
```diff
@@ -250,14 +230,26 @@
...
-        editing-message               (:editing-message metadata)
-        input-text-with-mentions      (mentions/check-mentions cofx input-text)]
-    (rf/merge cofx
-              (if editing-message
-                (send-edited-message input-text-with-mentions editing-message)
-                (send-messages input-text-with-mentions current-chat-id))
-              (mentions/clear-mentions)
-              (mentions/clear-cursor))))
+        editing-message               (:editing-message metadata)]
+    {:json-rpc/call [{:method     "wakuext_chatMentionCheckMentions"
```
Comment (flexsurfer) [inline src/status_im/chat/models/input.cljs]: probably we could keep this for the first iteration, but later this should be done inside `send-messages` so we shouldn't care about mentions in react, we just send-message and status-go check mentions itself

**Review by flexsurfer — APPROVED** [review]
awesome job, left a few comments

File: src/status_im/ui2/screens/chat/composer/input.cljs (hunk 1 above)
Comment (qfrank) [reply src/status_im/ui2/screens/chat/composer/input.cljs]: noob question .. can you explain what is "this issue"? 🙂 @flexsurfer

Comment (flexsurfer) [reply src/status_im/ui2/screens/chat/composer/input.cljs]: `type` was missed because `type` resolved as clojure.core function

Comment (flexsurfer) [reply src/status_im/ui2/screens/chat/composer/input.cljs]: similar happens often with `name` for example

Comment (qfrank) [reply src/status_im/ui2/screens/chat/composer/input.cljs]: i c. thank you

**Review by ilmotta — APPROVED** [review]
:clap: :clap:

**Comment by VolodLytvynenko** [comment]
Hi @qfrank thank you for PR. Here is some found issues:

## ISSUE 1: The mentions are not suggested after some special symbols or Cyrillic are present in the textbox

#### Steps to reproduce:
1. Go to any chat
2. Open the textbox
3. Paste text which contains special symbols or Cyrillic (Example: 'Привет' or ' "they" are watching us ')
4. Try to mention someone using '@'

#### Actual result:
The mentions are not suggested if some special symbols like  (", £) or Cyrillic are present in the textbox

https://user-images.githubusercontent.com/52490791/230097219-bc175eff-0635-4f79-8316-26db01fb95b5.mp4


#### Expected result:
The mentions are suggested

**Comment by VolodLytvynenko** [comment]
## ISSUE 2: [IOS] The suggestion with mentions is shown without inputted '@' in communities if initially, the textbox is fulfilled with mentions and the device comes back from the background mode

#### Preconditions:
User_A, User_B, and User_C are in the closed community
(Or use already a prepared user  with the community a seed phrase:
cause solution employ acquire only bacon drink boil blue dolphin welcome nation)

#### Steps to reproduce:
1. Go to the community
2. Open the textbox
3. Type some text and mention a few users (don't send this text)
4. Go to the background
5. Remove data into the textbox
6. Reopen the textbox

#### Actual result:
The suggestion with mentions is shown even if '@' is not entered

https://user-images.githubusercontent.com/52490791/230126463-9f2b495c-7b82-4a2f-ba3e-3ab6a7d10c6c.mp4

image


#### Expected result:
The suggestion with mentions is shown only after '@' is entered

#### Device:
- Real device: iPhone 11 Pro Max, IOS 16

**Comment by VolodLytvynenko** [comment]
## ISSUE 3: The keyboard is closed after the app comes back from the background
#### Steps to reproduce:
1. Go to any chat
2. Open the keyboard
3. Type some data into the textbox
4. Go to the background
5. Come back from the background

#### Actual result:
The keyboard is closed

https://user-images.githubusercontent.com/52490791/230128212-300f0fda-faf6-43e8-90af-d2cd2234e73b.mp4


#### Expected result:
The keyboard is still shown

#### Device:
- Real device: iPhone 11 Pro Max, IOS 16
- Real device: Huawei p20 light, android 9

**Comment by qfrank** [comment]
Hi @VladimrLitvinenko , thanks for reporting. issue 1 and 2 should be fixed now, pls retry, issue 3 should not be relate to this PR

**Comment by VolodLytvynenko** [comment]
hi, @qfrank issues 1,2 are fixed. Apologies for issue 3,  you are right, this issue is not related to the current PR. Reproduced it on the latest nightly

One additional issue is found:
## ISSUE 4: The mention is not shown as resolved if before the current mention located some special symbols or Cyrillic

#### Actual result:
The mention is not resolved if special symbols or cyrillic is used at the beginning of the message

https://user-images.githubusercontent.com/52490791/230392049-7601e2cc-cb0a-47c1-8a01-3b2ecb609e07.mp4


#### Expected result:
The mention is resolved
image

**Comment by qfrank** [comment]
> hi, @qfrank issues 1,2 are fixed. Apologies for issue 3, you are right, this issue is not related to the current PR. Reproduced it on the latest nightly
> 
> One additional issue is found:
> 
> ## ISSUE 4: The mention is not shown as resolved if before the current mention located some special symbols or Cyrillic
> #### Actual result:
> The mention is not resolved if special symbols or cyrillic is used at the beginning of the message
> 
> #### Expected result:
> The mention is resolved image

thanks for your feedback, it should be fixed now @VladimrLitvinenko

**Comment by VolodLytvynenko** [comment]
hi @qfrank PR is tested and ready to be merged. No issues from my side

**Pull request merged**